### ISSUE TITLE
Exception

### ISSUE THREAD

**Issue by Shadow505** [nWidart/laravel-modules#1081]
Hello,

I run an installation with about 10 modules. Unfortunately I am experiencing issues with the plugin. From time to time an exception like this occurs:

```
ErrorException: file_get_contents(): read of 8192 bytes failed with errno=22 Invalid argument in /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php:49
Stack trace:
#0 [internal function]: Illuminate\Foundation\Bootstrap\HandleExceptions->handleError()
#1 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php(49): file_get_contents()
#2 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Json.php(112): Illuminate\Filesystem\Filesystem->get()
#3 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Json.php(122): Nwidart\Modules\Json->getContents()
#4 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Json.php(41): Nwidart\Modules\Json->getAttributes()
#5 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Module.php(227): Nwidart\Modules\Json->__construct()
#6 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Support/helpers.php(521): Nwidart\Modules\Module->Nwidart\Modules\{closure}()
#7 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Support/Arr.php(304): value()
#8 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Module.php(228): Illuminate\Support\Arr::get()
#9 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Module.php(241): Nwidart\Modules\Module->json()
#10 /home/vagrant/project/vendor/nwidart/laravel-modules/src/FileRepository.php(293): Nwidart\Modules\Module->get()
#11 [internal function]: Nwidart\Modules\FileRepository->Nwidart\Modules\{closure}()
#12 /home/vagrant/project/vendor/nwidart/laravel-modules/src/FileRepository.php(302): uasort()
#13 /home/vagrant/project/vendor/nwidart/laravel-modules/src/FileRepository.php(330): Nwidart\Modules\FileRepository->getOrdered()
#14 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Providers/BootstrapServiceProvider.php(15): Nwidart\Modules\FileRepository->boot()
#15 [internal function]: Nwidart\Modules\Providers\BootstrapServiceProvider->boot()
#16 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): call_user_func_array()
#17 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Container/Util.php(37): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#18 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\Container\Util::unwrapIfClosure()
#19 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\Container\BoundMethod::callBoundMethod()
#20 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Container/Container.php(590): Illuminate\Container\BoundMethod::call()
#21 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(856): Illuminate\Container\Container->call()
#22 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(839): Illuminate\Foundation\Application->bootProvider()
#23 [internal function]: Illuminate\Foundation\Application->Illuminate\Foundation\{closure}()
#24 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(840): array_walk()
#25 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/BootProviders.php(17): Illuminate\Foundation\Application->boot()
#26 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(219): Illuminate\Foundation\Bootstrap\BootProviders->bootstrap()
#27 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(156): Illuminate\Foundation\Application->bootstrapWith()
#28 /home/vagrant/project/app/Http/Kernel.php(91): Illuminate\Foundation\Http\Kernel->bootstrap()
#29 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(140): App\Http\Kernel->bootstrap()
#30 /home/vagrant/project/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter()
#31 /home/vagrant/project/public/index.php(55): Illuminate\Foundation\Http\Kernel->handle()
```

It seems that  the chance that this occurs increases when the amount of requests sent to the website increases. Any idea where that comes from?

Thanks and best regards

**Comment by nWidart** [comment]
Did you check at `#3 /home/vagrant/project/vendor/nwidart/laravel-modules/src/Json.php(122): Nwidart\Modules\Json->getContents()` to see what's happening?

Maybe you don't have the module.json files or something similar.

**Comment by Shadow505** [comment]
Hello,

I am not sure what exactly I should be looking for. I added a catch() and logged the file path that is being returned from `$this->getPath()`. As expected, it is correct. As mentioned, the issues does not occur for every request. As you can see in the exception it says:

` file_get_contents(): read of 8192 bytes failed with errno=22 Invalid argument `

The issues is also not limited to a specific module. I am having trouble finding any additional information that might be useful.. 

Thanks in advance for your help!

**Comment by Shadow505** [comment]
I have set up ApacheBenchmark and ran it for a while with the somewhat highest settings that the server can handle without producing errors like "Bad gateway", etc, or in other word settings that are as high as possible but as low as needed for them to get processed properly by php-fpm. 

After a while I got the exception again. There is definitely something wrong during initializing the modules/loading there config. I also stumbled accross posts like this:
https://bugs.php.net/bug.php?id=78482

I am not sure if it is a coincidence but I am too running PHP 7.4.

**Comment by nWidart** [comment]
Hi,

What I meant was, put a breakpoint / `dd()` call in the `getContents` method and see what happens there.
Judging from your link it looks like your files might not have write permissions, so that's something to look at too.

**Comment by Shadow505** [comment]
Hi,

what exactly should be dd()'d? 

No, as mentioned that is not the case, otherwise the issue would occur every time, but it does not. I am not sure how exactly the process for loading the module data is, but given that the issue seems to occur especially if there are many requests being sent to the server, something might prevent the file from being access properly if it is currently already being used by another php process. But that's only speculation based on the information I was able to gather about the issue.

**Comment by nWidart** [comment]
There isn't much to dump, only 2 things. The path and the get on the filesystem.
Please check your permissions anyway.

**Comment by Shadow505** [comment]
I already mentioned that I have checked the output of `$this->getPath()` inside of the `getContents()` method. It is correct.

I don't think it is possible to dd() the `$this->filesystem->get()` because that's pretty much the call that causes the exception.

**Comment by nWidart** [comment]
Well then there's not much I can do with this information.
As you can see this method delegates to laravel filesystem so you could try to post an issue there.

**Comment by Shadow505** [comment]
If there is any other specific information that I could provide please let me know. unfortunately I don't think that's the case.. The call to the laravel filesystem is the last one, - the file might have become (temporarily) inaccessible due to the steps that your package takes.

**Comment by nWidart** [comment]
If that were the case, you wouldn't be the first one with this issue.
From google it seems pretty clear it could be a permissions issue. (yes even if not every request).

If you read the code you will so not much happen with this file, too.

**Comment by Shadow505** [comment]
Not to be rude but "If that were the case, you wouldn't be the first one with this issue." is really not a very problem-solving argument. There is always going to be a first one. One could just as well argue that the other users of the package are not having enough requests to the website to cause the issue.

I have not found a post that indicates a permission issue in the sense of wrong access permissions. They all boil down to the file being inaccessible, that could very well happen as described above - and it is in fact the only thing that seems to be logical to me. The file permissions are definitely correct.

**Comment by nWidart** [comment]
If you have a special use case, and there's a new issue, that's great, and please do contribute to improve the project. But here you're not providing any information.
There isn't much I can do especially since it's only calling laravel.

Example about permissions: https://github.com/php/php-src/blob/77a0fa101ed59f05ea07b03adb1cb66962fc45e7/UPGRADING#L149-L150

**Comment by Shadow505** [comment]
What is a "special" use case? Any use case that produces the error described above?

I have already seen the linked post. Nowhere it says that it's limited to incorrect file permissions. It says

>  trying to write to a read-only file resource

which will for example also occur if you open a file handler in one process and try to write to the same file in another one. That's exactly what I have described above.

> But here you're not providing any information.

I have asked you for the specific additional information that I could provide, but have not received instructions to that.

I have provided as much useful information as I can think of and have provided instructions to somewhat reproduce the issue (with ApacheBench, PHP 7.4).

You're just assuming that there is nothing wrong with the package, I am not sure why..

**Comment by nWidart** [comment]
`read-only file resource` could also be interpreted as permissions...

You were the one mentioning high requests, now you say it's not the case.

I'll close this as there doesn't seem to be anything conclusive.

**Comment by Shadow505** [comment]
> could also be interpreted as permissions...

Yes, could. But I have mentioned numerous times that this is NOT the case. That leaves pretty much only what I have described above.

> You were the one mentioning high requests, now you say it's not the case.

I was not. I said:

`It seems that the chance that this occurs increases when the amount of requests sent to the website increases. `

And then said that I have set up ApacheBench to test with the highest amount of requests that the server can handle. The server is a 5$ digital ocean droplet.

> I'll close this as there doesn't seem to be anything conclusive.

Impressive observation. Its not like I provided a full stacktrace, as much information as I can think of, steps to reproduce the issue and also have asked multiple times for instructions for getting any additional specific information that I could retrieve.
